src/storage/mod.rs

```diff
@@ -129,6 +241,150 @@ impl Storage {
         Ok(blob)
     }
 
+    pub(super) fn get_range(
+        &self,
+        path: &str,
+        max_size: usize,
+        range: FileRange,
+        compression: Option<CompressionAlgorithm>,
+    ) -> Result<Blob, Error> {
+        let mut blob = match &self.backend {
+            StorageBackend::Database(db) => db.get(path, max_size, Some(range)),
+            StorageBackend::S3(s3) => s3.get(path, max_size, Some(range)),
+        }?;
+        // `compression` represents the compression of the file-stream inside the archive.
+        // We don't compress the whole archive, so the encoding of the archive's blob is irrelevant
+        // here.
+        if let Some(alg) = compression {
+            blob.content = decompress(blob.content.as_slice(), alg, max_size)?;
+            blob.compression = None;
+        }
+        Ok(blob)
+    }
+
+    fn get_index_for(&self, archive_path: &str) -> Result<archive_index::Index, Error> {
+        // remote/folder/and/x.zip.index
+        let remote_index_path = format!("{}.index", archive_path);
+        let local_index_path = self.local_archive_cache_path.join(&remote_index_path);
+
+        if local_index_path.exists() {
+            let mut file = fs::File::open(local_index_path)?;
+            archive_index::Index::load(&mut file)
+        } else {
+            let index_content = self.get(&remote_index_path, std::usize::MAX)?.content;
+
+            fs::create_dir_all(
+                local_index_path
+                    .parent()
+                    .ok_or_else(|| err_msg("index path without parent"))?,
+            )?;
+            let mut file = fs::File::create(&local_index_path)?;
+            file.write_all(&index_content)?;
+
+            archive_index::Index::load(&mut &index_content[..])
+        }
+    }
+
+    pub(crate) fn get_from_archive(
+        &self,
+        archive_path: &str,
+        path: &str,
+        max_size: usize,
+    ) -> Result<Blob, Error> {
+        let index = self.get_index_for(archive_path)?;
+        let info = index.find_file(path)?;
+
+        let blob = self.get_range(
+            archive_path,
+            max_size,
+            info.range(),
+            Some(info.compression()),
+        )?;
+
+        Ok(Blob {
+            path: format!("{}/{}", archive_path, path),
+            mime: detect_mime(&path).into(),
+            date_updated: blob.date_updated,
+            content: blob.content,
+            compression: None,
```

Comment:
Shouldn't this be `blob.compression`? If you're sure this has already been decompressed, can you add an assert for that?

Reply:
since in `get_from_archive` we always call `get_range` with `Some(compression_alg)`, the resulting blob will always be decompressed. 

I'll add an assert.